students/Dustin_L/lesson07/html_render.py

```diff
@@ -0,0 +1,180 @@
+        if self.attrs:
+            attrs = ' ' + ' '.join(f'{k}="{v}"' for k, v in self.attrs.items())
+        else:
+            attrs = ''
+
+        return attrs
+
+
+class HtmlElement(Element):
+    """HTML type Element"""
+    tag = 'html'
+
+    def render(self, file_out, cur_ind=''):
+        """Renders the tag and strings from the element content.
+
+        Args:
+            file_out (file): Writable file-like object to recieve rendered data.
+            cur_ind (str, optional): Defaults to ''. Indentation of current element.
+        """
+        file_out.write(f'{cur_ind}<!DOCTYPE html>\n')
+        super().render(file_out, cur_ind)
+
+
+class BodyElement(Element):
+    """Body type Element"""
+    tag = 'body'
+
+
+class ParagraphElement(Element):
+    """Paragraph type Element"""
+    tag = 'p'
+
+
+class HeadElement(Element):
+    """Head type Element"""
+    tag = 'head'
+
+
+class OneLineElement(Element):
+    """One line type Element. Overrides Element.render()"""
+    def render(self, file_out, cur_ind=''):
+        """Renders the tag and strings from the element content onto one line.
+
+        Args:
+            file_out (file): Writable file-like object to recieve rendered data.
+            cur_ind (str, optional): Defaults to ''. Indentation of current element.
+        """
+        file_out.write(f'{cur_ind}<{self.tag}{self.fmt_attrs()}> ')
+
+        for item in self.content:
+            if hasattr(item, 'render'):
+                item.render(file_out, cur_ind + self.indent)
+            else:
+                file_out.write(str(item) + ' ')
+
+        file_out.write(f'</{self.tag}>\n')
+
+
+class TitleElement(OneLineElement):
+    """Title type Element"""
+    tag = 'title'
+
+
+class SelfClosingElement(Element):
+    """Self-Closing type Element"""
+    def __init__(self, content=None, **attrs):
+        super().__init__(content=None, **attrs)
+
+    # Overwrite content attribute to prevent use
+    @property
+    def content(self):
+        return None
+
+    @content.setter
+    def content(self, value):
+        pass
+
+    def render(self, file_out, cur_ind=''):
+        """Renders the tag and attrs as self-closing element.
+
+        Args:
+            file_out (file): Writable file-like object to recieve rendered data.
+            cur_ind (str, optional): Defaults to ''. Indentation of current element.
+        """
+        file_out.write(f'{cur_ind}<{self.tag}{self.fmt_attrs()} />\n')
+
+
+class HrElement(SelfClosingElement):
+    """hr type self-closing element"""
+    tag = 'hr'
+
+
+class BrElement(SelfClosingElement):
+    """br type self-closing element"""
+    tag = 'br'
+
+
+# NOTE: Shouldn't this sublclass maintain the same init params as Element
+#       in order to adhere to the Liskov Substitution principle?
+class AnchorElement(OneLineElement):
+    """Anchor link type element"""
+    tag = 'a'
+
+    def __init__(self, link, content):
+        super().__init__(content, href=link)
+
```

Comment:
I'm not quite sure if this gets to your question and maybe @natasha-aleksandrova has some insight into the application of LSP in dynamically typed languages and whether that applies here but a common pattern in Python is to use *args and **kwargs when overriding and/or extending the `__init__ ` method in a subclass, so this could be rewritten as:
```
    def __init__(self, link, *args, **attrs):
        attrs['href'] = link
        super().__init__(*args, **attrs)
```